src/ngx_http_lua_api.c

```diff
@@ -194,13 +198,12 @@ ngx_http_lua_shared_memory_init(ngx_shm_zone_t *shm_zone, void *data)
 
     lmcf->shm_zones_inited++;
 
-    if (lmcf->shm_zones_inited == lmcf->shm_zones->nelts
-        && lmcf->init_handler)
+    if (lmcf->shm_zones_inited == lmcf->shm_zones->nelts)
     {
         saved_cycle = ngx_cycle;
         ngx_cycle = ctx->cycle;
 
-        rc = lmcf->init_handler(ctx->log, lmcf, lmcf->lua);
+        rc = ngx_http_lua_delay_init_handler();
```

Comment:
@agentzh The 3rd-part won't relay on the ngx_lua's own `lua_shared_dict`, they can use `ngx_http_lua_delay_init_handler` in their own module like [this](https://github.com/openresty/lua-nginx-module/pull/1050/files#diff-b4a8646df8a3d1d284dc4ab1916cbfbbR217), and `init_by_lua` will be delayed. 
But the disadvantage is that the third party module may need to maintain a `ngx_cycle`.